File: Dynamic-Programming/139. Word Break/wordBreak.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
 var wordBreak = function(s, wordDict) {
    let n = s.length;
    let dp = new Array(n + 1).fill(false);
    dp[0] = true;
    
    let maxLen;
    for (const word in wordDict) {
        maxLen = Math.max(maxLen, word.length);
    }
    
    for (let i = 1; i <= n; i++) {
        for (let j = 0; j < i; j++) {
            if (i - j > maxLen) {
                continue;
            }
            if (dp[j] && wordDict.includes(s.substring(j, i))) {
                dp[i] = true;
                break;
            }
        }
    }
    
    return dp[n];
};

File: Dynamic-Programming/139. Word Break/wordBreak.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        n = len(s)
        dp = [False for _ in range(n + 1)]
        dp[0] = True
        
        max_len = 0
        for word in wordDict:
            max_len = max(max_len, len(word))

        for i in range(1, n + 1):
            for j in range(0, i):
                if i - j > max_len:
                    continue
                if dp[j] and s[j:i] in wordDict:
                    dp[i] = True
                    break
        
        return dp[-1]